enhancements/20220413-zero-downtime-upgrade.md

```diff
@@ -0,0 +1,202 @@
+
+The upgrade controller runs a job with `harvester-upgrade:<new_version>` image to upgrade system services, including:
+- Rancher and its dependencies.
+- Harvester cluster-repo.
+- Harvester chart.
+- Monitoring chart.
+- Any other fixes or migrations from previous versions.
+
+**Phase 4**: Upgrade nodes
+
+The upgrade controller informs Rancher to upgrade the RKE2 runtimes on each node. We use the [pre-drain and post-drain hooks feature](https://github.com/rancher/rancher/issues/35464) to perform additional works. The following is what happens when upgrading a node.
+- Rancher starts to upgrade a node and notifies Harvester to run the pre-drain job.
+- The Harvester upgrade controller creates a pre-drain job with `harvester-upgrade:<new_version>` image to:
+  - Live-migrates VMs out of the node.
+  - Shutdown non-migratable VMs.
+- The Harvester controller notifies Rancher that the pre-drain job is done.
+- Rancher starts to upgrade RKE2 server or agent on the node.
+- Rancher notifies Harvester to run the post-drain job.
+- The Harvester controller creates a post-drain job with `harvester-upgrade:<new_version>` image to:
+  - Wait for RKE2 runtime to upgrade.
+  - Upgrade the node's OS with `cos` script.
+  - Clean up unused images.
+  - Reboot the node.
+- When the node is back, the Harvester controller notifies Rancher that the post-hook is done. Rancher starts to upgrade the next node.
+
+**Phase 5**: Clean-up
+
+The Harvester controller will delete the upgrade repository VM and its image.
+
+### Test plan
+
+- Spawn Harvester clusters with a previous release. Ideally, the test should cover:
+  - Single node Harvester deployment.
+  - 2-nodes Harvester cluster.
+  - 3-nodes (or more) Harvester cluster.
+
+- Save the Harvester ISO to a local http server and calculate its SHA-512 checksum.
+- Create a `Version` CR:
+
+    ```
+    apiVersion: harvesterhci.io/v1beta1
+    kind: Version
+    metadata:
+      name: v1.0.2      # <-- need to be newer than current version
+      namespace: harvester-system
+    spec:
+      isoChecksum: <replace me>
+      isoURL: <replace me>
+      releaseDate: '20220408'
+      tags:
+      - dev
+    ```
+- Click the "upgrade" button in the GUI to start the upgrade.
+- Harvester ISO contains components' versions, verify them after the upgrade:
+
+    ```
+    $ cat <iso_mnt_path>/harvester-release.yaml
+    harvester: v1.0.1
+    harvesterChart: 1.0.1
+    os: Harvester v1.0.1
+    kubernetes: v1.21.11+rke2r1
+    rancher: v2.6.4-harvester3
+    monitoringChart: 100.1.0+up19.0.3
+    kubevirt: 0.49.0-2
+    rancherDependencies:
+      fleet:
+        chart: 100.0.3+up0.3.9
+        app: 0.3.9
+      fleet-crd:
+        chart: 100.0.3+up0.3.9
+        app: 0.3.9
+      rancher-webhook:
+        chart: 1.0.4+up0.2.5
+        app: 0.2.5
+    ```
+
+### Upgrade strategy
+
+No.
+
+## Note
+
+- With the current approach, the last-upgraded node contains no VMs (VMs are live-migrated out of the node).
```

Comment:
I feel like this is like a limitation or caveat. Can we state it more prominently? Something like

```suggestion
### Limitations

- With this proposed approach, the last-upgraded node contains no VMs (VMs are live-migrated out of the node).
```

Reply:
Sound good!